### PR TITLE
Ignore metric weights in layers

### PR DESCRIPTION
**Pull request by lgeiger**
This will ignore the weights added by metrics when calling `layer.weights`. This has a few benefits:
- Weights from layer metrics will not be saved when calling `model`. This means that:
  - checkpoints are compatible between different versions of `larq` and when disabling metrics on the user side. I discovered this when trying to reload an old weights file in `larq-zoo` that didn't include the metrics.
  - checkpoints are smaller since they don't include unnecessary items
- We don't need to handle the ignoring in the `model.summary` code (which is already quite bloated)
- In general interacting with `layer.weights` will be easier since weights related to metrics are hidden for users

### CODE REVIEW

File: larq/layers.py [hunk 1]
```diff
@@ -56,6 +56,14 @@ def build(self, input_shape):
                     values_shape=self.kernel.shape, name=f"{self.name}/flip_ratio"
                 )
 
+    @property
+    def non_trainable_weights(self):
```
Comment (lgeiger) [inline larq/layers.py]: We really need to find a better way to share logic between layers. I'll do some refactoring next week. We should also checkout how TensorFlow is handling their automatic mixed precision [here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/keras/mixed_precision/experimental/autocast_variable.py) to maybe gain some inspiration.

Comment (lgeiger) [inline larq/layers.py]: @koenhelwegen Do you think messing with `non_trainable_weights` could have any unexpected side effects?

I also tried setting `non_trainable_weights = []` on the metrics layer directly, but that doesn't work since initialisation wouldn't be handled correctly.

Comment (koenhelwegen) [reply larq/layers.py]: I have a hard time judging if this is save. But I don't really understand why we have to mess with this in the first place. TF Keras metrics are also stateful, so is there not a more standard way of dealing with this?

Comment (lgeiger) [reply larq/layers.py]: > TF Keras metrics are also stateful, so is there not a more standard way of dealing with this?

TF Keras metrics are added to the model. We add our metrics to the layer since we want to observe changes in the forward pass. Underneath a metric is a normal Keras layer and therefor adding weights will add them to the layer as well.
`layer.weights` internally calls `layers.non_trainable_weights` and `layers.trainable_weights` and will also get all the weights from its child layers (which is important if you have layers that bundle multiple layers). I checked, that the metrics are correctly updated during training in TF 1.14 and TF 2. It seams that we can safely make this change without messing things up.

I'll open an issue on the TensorFlow repo on Monday to see if this behaviour is intended or an oversight. The `add_metrics` API is quite new so there aren't a lot of examples out there that use this on a layer level.

Comment (koenhelwegen) [reply larq/layers.py]: okay, thanks! I will merge it and if we find a better way we can change it later
